client/python/openlineage/client/transport/async_http.py

```diff
@@ -0,0 +1,628 @@
+            return start_event_id in self.events
+
+    def _all_processed(self) -> bool:
+        with self.event_lock:
+            return len(self.events) == 0
+
+    def emit(self, event: Event) -> None:
+        event_str = Serde.to_json(event)
+
+        run_id = None
+        event_type = None
+
+        if isinstance(event, (RunEvent, RunEventV2)):
+            event_type = str(event.eventType.value)
+            event_id = event.run.runId + "-" + event_type
+            run_id = event.run.runId
+
+            # We don't want to emit terminal events until we have a successful START event
+            # Only queue as pending if START has been scheduled but not yet successful
+            # If START was never scheduled, process COMPLETE normally (fall through)
+            if event_type in ["COMPLETE", "FAIL", "ABORT"]:
+                if self._is_start_in_progress(run_id):
+                    body, headers = self._prepare_request(event_str)
+                    request = Request(
+                        event_id=event_id, run_id=run_id, event_type=event_type, body=body, headers=headers
+                    )
+                    self._add_event(request)
+                    self._add_pending_completion_event(request)
+
+                    log.debug(
+                        "Queued completion event %s for run %s, waiting for START event",
+                        event_type,
+                        run_id,
+                    )
+                    return
+        else:
+            event_id = hashlib.md5(event_str.encode("utf-8")).hexdigest()
+
+        body, headers = self._prepare_request(event_str)
+        request = Request(event_id=event_id, run_id=run_id, event_type=event_type, body=body, headers=headers)
+        self._add_event(request)
+
+        # Wait for space in the regular capacity portion of the queue
+        while self.event_queue.qsize() >= self.configured_queue_size:
+            time.sleep(0.01)
+
+        # Add to queue - include run_id and event_type for START event tracking
+        self.event_queue.put(request)
+        log.debug(f"Queued event with event_id {event_id} for async emission to {self.config.url}")
+
+    def wait_for_completion(self, timeout: float = 30.0) -> bool:
+        # Block until all events are processed or timeout is reached.
+        log.debug("Waiting for events to be sent in wait_for_completion.")
+        start_time = time.time()
+        while time.time() - start_time < timeout:
+            if self._all_processed():
+                log.debug("All events were processed.")
+                return True
+
+            # Brief sleep before checking again
+            time.sleep(0.01)
+
+        return self._all_processed()
+
+    def get_stats(self) -> dict[str, int]:
+        return self.event_stats
+
+    def shutdown(self, wait: bool = True, timeout: float = 30.0) -> bool:
+        if wait:
+            result = self.wait_for_completion(timeout)
+            self.should_exit.set()
```

Comment:
should_exit should be moved to the beginning of the method, to notify threads about pending shutdown

Reply:
This being after `wait_for_completion` lets the transport process events from the queue - not only those actively being emitted. `should_exit` blocks the worker thread from getting next event from the queue.

I think with supplied timeout this behavior is desired, we can imagine some integration emitting event, then immediately calling shutdown with 10 second timeout - I would imagine this should try to emit the event.

Reply:
Ok, good point